src/rrt/Tree.hpp

```diff
@@ -254,23 +311,41 @@ class Tree {
     /**
      * Find the node int the tree closest to @state.  Pass in a double pointer
      * as the second argument to get the distance that the node is away from
-     * @state.
+     * @state. This method searches a k-d tree of the points to determine
      */
     Node<T>* nearest(const T& state, double* distanceOut = nullptr) {
-        double bestDistance = -1;
         Node<T>* best = nullptr;
 
-        for (Node<T>& other : _nodes) {
-            double dist = _stateSpace->distance(other.state(), state);
-            if (bestDistance < 0 || dist < bestDistance) {
-                bestDistance = dist;
-                best = &other;
-            }
+        // k-NN search (O(log(N)))
+        flann::Matrix<double> query;
+        if (NULL == _TToArray) {
+            query = flann::Matrix<double>((double*)&state, 1,
+                                          sizeof(state) / sizeof(0.0));
+        } else {
+            std::vector<double> data(_dimensions);
+            _TToArray(state, data.data());
+            query = flann::Matrix<double>(data.data(), 1,
+                                          sizeof(state) / sizeof(0.0));
         }
+        std::vector<int> i(query.rows);
+        flann::Matrix<int> indices(i.data(), query.rows, 1);
+        std::vector<double> d(query.rows);
+        flann::Matrix<double> dists(d.data(), query.rows, 1);
 
-        if (distanceOut) *distanceOut = bestDistance;
+        int n =
+            _kdtree.knnSearch(query, indices, dists, 1, flann::SearchParams());
+
+        if (distanceOut)
+            *distanceOut = _stateSpace->distance(state, best->state());
+
+        T point;
+        if (NULL == _arrayToT) {
+            point = (T)_kdtree.getPoint(indices[0][0]);
```

Comment:
I'll update the documentation for this. If this cast is not valid then the user is expected to pass in a function pointer for _arrayToT which is used instead of making this cast.